### ISSUE TITLE
Best Practices for Saving and Evaluating during Training? SAC Policy with HER [question]

### ISSUE THREAD

**Issue by evanatyourservice** [hill-a/stable-baselines#376]
Hello!

I've had some questions about the best practices for saving and evaluating during training, and haven't gotten completely clear answers from the documentation or other issues.

I am working in Google Colab with this HER+SAC setup:

```python
from stable_baselines import HER, SAC
from stable_baselines.sac.policies import FeedForwardPolicy

env = MyEnv(myconfig, '/content/train/')

class CustomSACPolicy(FeedForwardPolicy):
    def __init__(self, *args, **kwargs):
        super(CustomSACPolicy, self).__init__(*args, **kwargs,
                                           layers=[400, 300],
                                           layer_norm=False,
                                           feature_extraction="mlp")

model = HER(CustomSACPolicy, env, SAC, n_sampled_goal=4,
            goal_selection_strategy='future',
            verbose=1, buffer_size=int(1e6),
            learning_rate=1e-3,
            random_exploration=0.01,
            gamma=0.99, batch_size=512,
            learning_starts=5000)
```

When I do:

```python
iteration_length = 30000
iterations = 100

eval_env = MyEnv(myconfig, '/content/test/', eval=True)

for _ in range(iterations):
    print('training')
    model.learn(iteration_length, log_interval=10)
    print('saving')
    model.save('/content/drive/My Drive/Colab Notebooks/sac_training')
    print('evaluating')
    model.set_env(eval_env)
    obs = eval_env.reset()
    episode_reward = 0.0
    rewards = []
    for _ in range(myconfig['episode_len']*4+5):
        action, _ = model.predict(obs)
        obs, reward, done, info = eval_env.step(action)
        episode_reward += reward
        if done:
            rewards += [episode_reward]
            print("Reward =", round(episode_reward, 2), "Info:", info)
            episode_reward = 0.0
            obs = eval_env.reset()
    print("Mean Reward = ", round(np.mean(rewards), 2))
    model.set_env(env)
```
It seems some things are reset after saving and evaluating every time, but some things are not, perhaps the things mentioned in #301. Is there a safe way to save and evaluate during training while preserving the training process, or is this generally not yet a feature, with the complete training state preserved only if training is uninterrupted?

Can a callback be safely used to save and/or evaluate during training?
Can the environment be safely changed mid-training for evaluating on different data than training data?
If there is no completely safe way, perhaps the safest way is to just increase the length of each iteration to take advantage of the current training state as long as possible before saving and evaluating.

**Comment by evanatyourservice** [comment]
After looking at more training results it looks as if training remains stable enough after stopping and starting, even if the model is deleted and completely reloaded from file. In this case I think the brief inconsistencies in stats between training iterations I’m seeing will be solved by #301

If there is a better way of doing what I have above I’d still be interested though! I saw in another issue that rendering with callback is possible if it is every step and not every n steps, but can the model be saved in a callback as well?

**Comment by evanatyourservice** [comment]
Actually I think the discrepancy is mostly because every time model.learn is called it goes through learning_starts=x again.

**Comment by araffin** [comment]
Hello,

> Is there a safe way to save and evaluate during training while preserving the training process, or is this generally not yet a feature, with the complete training state preserved only if training is uninterrupted?


The issue #301 only applies when the model is loaded.
Doing it the way you are doing has no normally no issue.


> Can the environment be safely changed mid-training for evaluating on different data than training data?

You don't need to do `set_env` to use a test env, you can directly call the predict method.
This has the advantage of not changing the state of the training env.

>Can a callback be safely used to save and/or evaluate during training?
>If there is a better way of doing what I have above I’d still be interested though! I saw in another issue that rendering with callback is possible if it is every step and not every n steps, but can the model be saved in a callback as well?

So to answer your main question, there are two main way of evaluating the model, you can either:

1. Do it the way you do it, but you should not change the training env and only use the predict method on the test env. 

2. Use a callback, you have an example of how yo do it in the [rl zoo](https://github.com/araffin/rl-baselines-zoo/blob/master/utils/hyperparams_opt.py#L79) The issue with the callback is that it is called only after n steps and it depends on the algorithm.

To save the model in the callback, you have an example in the [documentation](https://stable-baselines.readthedocs.io/en/master/guide/examples.html#using-callback-monitoring-training)

Last thing: be careful when using normalization wrapper, you need to keep the training and test env synced.

**Comment by evanatyourservice** [comment]
Ah thank you so much, this was a lot of help. I don't think I read the documentation closely enough! Can you elaborate on keeping the testing and training environments synced when using the layer_norm? What exactly do I need to make sure is the same between them?

**Comment by araffin** [comment]
> Can you elaborate on keeping the testing and training environments synced when using the layer_norm?

I am talking about VecNormalize, and the example in the rl zoo handles that.

**Comment by evanatyourservice** [comment]
Ah okay, thank you!

**Comment by evanatyourservice** [comment]
One quick question @araffin if you don't mind, the HER example with SAC uses 'MlpPolicy' in quotes as the policy... is this pulling from the SAC policies, or the common policies? Should I use the common policies in quotes when using SAC with HER, or use the SAC policies in quotes, or import a SAC policy and pass it in as is? I got an error when trying to evaluate an HER model using a custom SAC model and am wondering if it is because of the SAC model.

eval_env = DeribitTrainEnv(deribittrainconfig, '/content/test/', eval=True)
#eval_env = DummyVecEnv([lambda: eval_env])

```
model = HER.load('/content/drive/My Drive/Colab Notebooks/stable_baselines/sac_training', env=eval_env)

print('evaluating')
obs = eval_env.reset()
episode_reward = 0.0
rewards = []
for _ in range(config['episode_len']*10+5):
    action, _ = model.predict(obs)
    obs, reward, done, info = eval_env.step(action)
    episode_reward += reward
    if done:
        rewards += [episode_reward]
        print("Reward =", episode_reward, "Info:", info)
        episode_reward = 0.0
        obs = eval_env.reset()
print("Mean Reward = ", round(np.mean(rewards), 2))
```
Error:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-11-e1e7549c1754> in <module>()
      3 
      4 #
----> 5 model = HER.load('/content/drive/My Drive/Colab Notebooks/stable_baselines/sac_training', env=eval_env)
      6 
      7 print('evaluating')

3 frames
<ipython-input-2-d342d16ccd6a> in __init__(self, *args, **kwargs)
      8 class CustomSACPolicy(FeedForwardPolicy):
      9     def __init__(self, *args, **kwargs):
---> 10         super(CustomSACPolicy, self).__init__(*args, **kwargs,
     11                                            layers=[400, 300],
     12                                            layer_norm=True,

TypeError: super(type, obj): obj must be an instance or subtype of type
```

**Comment by araffin** [comment]
Please read the documentation about SAC. When you pass a string for the policy, it will find automatically the right one (in that case, MlpPolicy specific to SAC).

**Comment by evanatyourservice** [comment]
Any idea why I'm getting that error when I load the model that was trained on the custom policy?

**Comment by araffin** [comment]
Put your custom policy in a seperate file, you can take a look at the rl zoo or use net_kwargs.

**Comment by araffin** [comment]
Also in the documentation: https://stable-baselines.readthedocs.io/en/master/guide/custom_policy.html

"When loading a model with a custom policy, you must pass the custom policy explicitly when loading the model. (cf previous example)"

**Comment by evanatyourservice** [comment]
Thank you!
